File: java/JRM/2018098_A2/LinkedList.java
import java.util.*;

class LinkedList {
    static LNode  head;
    static int size = 0;
    LinkedList(int v) {
        size++;
        LNode new_node = new LNode(v); 

        if (head == null) 
        { 
            head = new LNode(v); 
            return; 
        } 
        new_node.next = null; 

        LNode last = head;  
        while (last.next != null) 
            last = last.next; 
        last.next = new_node; 
        return;
    }

    LinkedList(LNode new_node) {
        size++;
        if (head == null) 
        { 
            head = new_node;
            return; 
        } 
        new_node.next = null; 

        LNode last = head;  
        while (last.next != null) 
            last = last.next; 
        last.next = new_node; 
        return;
    }
    

    static class LNode {
        int key;
        LNode next;
        LNode(int k) {
            key = k;
            next = null;
        }
    }

    void swap(LNode x) {
        int a = x.key;
        int b = x.next.key;
        x.key = b;
        x.next.key = a;
    }
    void sort() {
        LNode temp = head;
        for(int i=0; i<size; i++) {
            for(int j=1; j<size; j++) {
                if (temp.key>temp.next.key) {
                    swap(temp);
                }
                temp = temp.next;
            }
            temp = head;
        }
    }

    void display() {
        LNode temp = head;
        while(temp!=null) {
            System.out.print(temp.key + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    void find() {
        LNode temp = head;
        System.out.print(temp.key + " ");
        while(temp.next!=null) {
            temp = temp.next;
        }
        System.out.println(temp.key);
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList(5);
        ll = new LinkedList(4);
        ll = new LinkedList(3);
        ll = new LinkedList(2);
        ll = new LinkedList(1);

        ll = new LinkedList(new LNode(10));
        ll = new LinkedList(new LNode(9));
        ll = new LinkedList(new LNode(8));
        ll = new LinkedList(new LNode(7));
        ll = new LinkedList(new LNode(6));

        ll.display();
        ll.sort();
        ll.display();
        ll.find();

    }
}

File: c++/ds/LinkedList.cpp
#include <iostream>
using namespace std;

struct Node
{
    int data;
    Node* next;
};

class LinkedList
{
public:
    Node* head;

    LinkedList() { head = NULL; }
    
    void push_back(int val)
    {
        Node* temp = head;
        Node* new_node = new Node();
        new_node->data = val;
        new_node->next = NULL;

        if (temp == NULL) {
            head = new_node;
            return;
        }

        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = new_node;
    }

    void push_front(int val)
    {
        Node* new_node = new Node();
        new_node->data = val;
        new_node->next = head;
        head = new_node;
    }

    void delete_node(int val)
    {
        Node* temp = head;
        if (temp->data == val) {
            Node *next = head->next;
            free(head);
            head = next;
            return;
        }
        while(temp->next != NULL && temp->next->data != val) {
            temp = temp->next;
        }
        if (temp->next != NULL) {
            Node* next = temp->next->next;
            free(temp->next);
            temp->next = next;
        }
    }

    void print()
    {
        Node* temp = head;
        while (temp != NULL) {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }
};

int main()
{
    LinkedList LL;
    LL.push_back(2);
    LL.push_back(5);
    LL.push_back(7);
    LL.push_back(9);
    LL.push_back(4);
    LL.print();
    LL.push_front(2);
    LL.print();
    LL.delete_node(7);
    LL.print();
}